daemon/migrator.c

```diff
@@ -0,0 +1,804 @@
+approval_uid(const approval_t *self)
+{
+    /* UID from approval file path */
+    return self->apr_uid;
+}
+
+static stringset_t *
+approval_permissions(const approval_t *self)
+{
+    /* Approved permissions */
+    return self->apr_permissions;
+}
+
+static const gchar *
+approval_organization(const approval_t *self)
+{
+    /* The same as OrganizationName in Sailjail section */
+    return self->apr_organization;
+}
+
+static const gchar *
+approval_application(const approval_t *self)
+{
+    /* The same as ApplicationName in Sailjail section */
+    return self->apr_application;
+}
+
+#if DEAD_CODE
+static const gchar *
+approval_exec(const approval_t *self)
+{
+    /* Binary path, does not contain arguments */
+    return self->apr_exec;
+}
+#endif
+
+static const migrator_t *
+approval_migrator(const approval_t *self)
+{
+    return self->apr_migrator;
+}
+
+static const control_t *
+approval_control(const approval_t *self)
+{
+    return migrator_control(approval_migrator(self));
+}
+
+/* ------------------------------------------------------------------------- *
+ * UTILITY
+ * ------------------------------------------------------------------------- */
+
+#if DEAD_CODE
+static bool
+exec_matches_binary(const gchar *exec, const gchar *binary_path)
+{
+    gchar **argv = NULL;
+    gint    argc = 0;
+    bool    rv   = false;
+
+    /* Note: This can deal only with Exec lines that contain the
+     *       application as the first argument, i.e. it can't deal with
+     *       Exec lines that contains sailjail or invoker first.
+     */
+    if( g_shell_parse_argv(exec, &argc, &argv, NULL) &&
+        argc >= 1 && 
+        !g_strcmp0(argv[0], binary_path) )
+        rv = true;
+
+    g_strfreev(argv);
+    return rv;
+}
+#endif
+
+static void
+remove_approval_path(const gchar *path)
+{
+    /* Remove path and any empty parent directories until HOMESCREEN_DATA_PATH */
+    if( unlink(path) == 0 ) {
+        gchar *dir_path = path_dirname(path);
+        while( may_remove_approval_path(dir_path) && rmdir(dir_path) == 0 ) {
+            gchar *temp = dir_path;
+            dir_path = path_dirname(temp);
+            g_free(temp);
+        }
+        log_debug("%s cleaned up to %s", path, dir_path);
+        g_free(dir_path);
+    }
+    else {
+        log_err("%s could not remove: %m", path);
+    }
+}
+
+static bool
+may_remove_approval_path(const gchar *dir_path)
+{
+    return strlen(dir_path) > strlen(HOMESCREEN_DATA_PATH);
+}
+
+static gchar *
+profile_from_approval_path(const gchar *path)
+{
+    const gchar *approval_path = strchr(path + strlen(HOMESCREEN_DATA_PATH), '/');
+    return path_dirname(approval_path);
+}
+
+static uid_t
+uid_from_approval_path(const gchar *path, const control_t *control)
+{
+    const gchar *temp = path + strlen(HOMESCREEN_DATA_PATH);
+    gint64       uid  = g_ascii_strtoll(temp, NULL, 10);
+    if( uid >= control_min_user(control) &&
+        uid <= control_max_user(control) )
+        return uid;
```

Comment:
nit: gint64 is likely to be wider than uid_t -> having a (uid_t) cast would signal that it is intentional.

Reply:
Added that while I rewrote the function around that new `without_leading_data_path()` helper.